lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+from dataclasses import dataclass
+from typing import Iterable, Optional, Tuple
+
+from sqlalchemy import func, select
+
+from lms.models import (
+    Assignment,
+    AssignmentGrouping,
+    AssignmentMembership,
+    Course,
+    Grouping,
+    LTIRole,
+    User,
+)
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
+
+
 class DigestService:
     """A service for generating "digests" (activity reports)."""
 
+    def __init__(self, db, h_api, mailchimp_service, sender):
+        self._db = db
+        self._h_api = h_api
+        self._mailchimp_service = mailchimp_service
+        self._sender = sender
+
     def send_instructor_email_digests(
         self, audience, updated_after, updated_before, override_to_email=None
     ):
         """Send instructor email digests for the given users and timeframe."""
+        annotations = self._h_api.get_annotations(
+            audience, updated_after, updated_before
+        )
+
+        # HAPI.get_annotations() returns an iterable.
+        # Turn it into a tuple because we need to iterate over it multiple times.
+        annotations = tuple(annotations)
+
+        context = DigestContext(self._db, audience, annotations)
```

Comment:
I renamed `DigestHelper` to `DigestContext` as I think that's a better name. And a new one is now constructed for each call to `send_instructor_email_digests()` because `DigestContext.__init__()` now takes the `audience` and `annotations` as arguments.